src/data-model-providers/codegen/CodegenDataModelProvider.cpp

```diff
@@ -43,203 +47,221 @@
 {
-    mOperation = operation;
-    mTarget    = path.mCommandId;
 
     CommandHandlerInterface * interface =
         CommandHandlerInterfaceRegistry::Instance().GetCommandHandler(path.mEndpointId, path.mClusterId);
-
-    if (interface == nullptr)
+    if (interface != nullptr)
     {
-        return std::nullopt; // no data: no interface
+        size_t commandCount = 0;
+
+        CHIP_ERROR err = interface->EnumerateAcceptedCommands(
+            path,
+            [](CommandId id, void * context) -> Loop {
+                *reinterpret_cast<size_t *>(context) += 1;
+                return Loop::Continue;
+            },
+            reinterpret_cast<void *>(&commandCount));
+
+        if (err == CHIP_NO_ERROR)
+        {
+            using EnumerationData = struct
+            {
+                ConcreteCommandPath commandPath;
+                DataModel::MetadataList<DataModel::AcceptedCommandEntry> acceptedCommandList;
+                CHIP_ERROR processingError;
+            };
+
+            EnumerationData enumerationData;
+            enumerationData.commandPath     = ConcreteCommandPath(path.mEndpointId, path.mClusterId, kInvalidCommandId);
+            enumerationData.processingError = CHIP_NO_ERROR;
+
+            ReturnErrorOnFailure(enumerationData.acceptedCommandList.reserve(commandCount));
+
+            ReturnErrorOnFailure(interface->EnumerateAcceptedCommands(
+                path,
+                [](CommandId commandId, void * context) -> Loop {
+                    auto input                    = reinterpret_cast<EnumerationData *>(context);
+                    input->commandPath.mCommandId = commandId;
+                    CHIP_ERROR appendError        = input->acceptedCommandList.Append(AcceptedCommandEntryFor(input->commandPath));
+                    if (appendError != CHIP_NO_ERROR)
+                    {
+                        input->processingError = appendError;
+                        return Loop::Break;
+                    }
+                    return Loop::Continue;
+                },
+                reinterpret_cast<void *>(&enumerationData)));
+            ReturnErrorOnFailure(enumerationData.processingError);
+
+            // the two invocations MUST return the same sizes.
+            VerifyOrReturnError(enumerationData.acceptedCommandList.size() == commandCount, CHIP_ERROR_INTERNAL);
+
+            result = std::move(enumerationData.acceptedCommandList);
+            return CHIP_NO_ERROR;
+        }
+        VerifyOrReturnError(err == CHIP_ERROR_NOT_IMPLEMENTED, err);
     }
 
-    CHIP_ERROR err = (interface->*mCallback)(path, HandlerCallbackFn, this);
-    if (err == CHIP_ERROR_NOT_IMPLEMENTED)
+    if ((serverCluster == nullptr) || (serverCluster->acceptedCommandList == nullptr))
     {
-        return std::nullopt; // no data provided by the interface
+        // No data if cluster is not valid.
```

Comment:
`serverCluster->acceptedCommandList` is valid, though.  This makes it sound like that's an invalid configuration?

It would be better if here we explicitly set `result` to "empty list"....

Reply:
Updated comment a bit, however I believe logic is still correct: I would like to eventually move to a "append only" method and not allow clearing of lists, so that I can easily multiplex by calling some `GetFoo()` on different providers in sequence.

So I intentionally do not have a `Clear/Invalidate` call on the metadata. The comment would say "nothing to add".

Reply:
OK, that makes sense.